File: solutions/go/ocr-numbers/1/ocr_numbers.go
package ocr

import "strings"

var (
    charMapping = []byte{
        ' ', '_', ' ',
        '|', '_', '|',
        '|', '_', '|',
        ' ', ' ', ' ',
    }
    digitPositions = []map[int]struct{}{
        {1: {}, 3: {}, 5: {}, 6: {}, 7: {}, 8: {}},
        {5: {}, 8: {}},
        {1: {}, 4: {}, 5: {}, 6: {}, 7: {}},
        {1: {}, 4: {}, 5: {}, 7: {}, 8: {}},
        {3: {}, 4: {}, 5: {}, 8: {}},
        {1: {}, 3: {}, 4: {}, 7: {}, 8: {}},
        {1: {}, 3: {}, 4: {}, 6: {}, 7: {}, 8: {}},
        {1: {}, 5: {}, 8: {}},
        {1: {}, 3: {}, 4: {}, 5: {}, 6: {}, 7: {}, 8: {}},
        {1: {}, 3: {}, 4: {}, 5: {}, 7: {}, 8: {}},
    }
    offsetMapping = [][2]int{
        {0, 0}, {0, 1}, {0, 2},
        {1, 0}, {1, 1}, {1, 2},
        {2, 0}, {2, 1}, {2, 2},
        {3, 0}, {3, 1}, {3, 2},
    }
)

func recognizeDigit(matrix []string, basei, basej int) rune {
    for digit, positions := range digitPositions {
        matchFound := false
        for i, offset := range offsetMapping {
            data := matrix[basei + offset[0]][basej + offset[1]]
            if _, ok := positions[i]; ok {
                if charMapping[i] != data {
                    matchFound = false
                    break
                }
                matchFound = true
            } else {
                if ' ' != data {
                    matchFound = false
                    break
                }
                matchFound = true
            }
        }
        if matchFound {
            return '0' + rune(digit)
        }
    }
    return '?'
}

func Recognize(input string) []string {
    matrix := strings.Split(strings.Trim(input, "\n"), "\n")
	nrows := len(matrix)
    ncols := len(matrix[0])
    nOutRows, nOutCols := nrows / 4, ncols / 3
    result := make([]string, nOutRows)
    var b strings.Builder
    for i := 0; i < nOutRows; i++ {
        b.Reset()
        for j := 0; j < nOutCols; j++ {
            basei, basej := 4 * i, 3 * j
            b.WriteRune(recognizeDigit(matrix, basei, basej))
        }
        result[i] = b.String()
    }
    return result
}

File: solutions/javascript/ocr-numbers/2/ocr-numbers.js
const charMapping = [
  ' ', '_', ' ',
  '|', '_', '|',
  '|', '_', '|',
  ' ', ' ', ' ',
]

const digitPositions = [
  new Set([1, 3, 5, 6, 7, 8]),
  new Set([5, 8]),
  new Set([1, 4, 5, 6, 7]),
  new Set([1, 4, 5, 7, 8]),
  new Set([3, 4, 5, 8]),
  new Set([1, 3, 4, 7, 8]),
  new Set([1, 3, 4, 6, 7, 8]),
  new Set([1, 5, 8]),
  new Set([1, 3, 4, 5, 6, 7, 8]),
  new Set([1, 3, 4, 5, 7, 8]),
]

const offsetMapping = [
  [0, 0], [0, 1], [0, 2],
  [1, 0], [1, 1], [1, 2],
  [2, 0], [2, 1], [2, 2],
  [3, 0], [3, 1], [3, 2],
]

function recognizeDigit(matrix, basei, basej) {
  for (const [digit, positions] of digitPositions.entries()) {
    let matchFound = false
    for (const [i, offset] of offsetMapping.entries()) {
      const data = matrix[basei + offset[0]][basej + offset[1]]
      if (positions.has(i)) {
        if (charMapping[i] !== data) {
          matchFound = false
          break
        }
        matchFound = true
      } else {
        if (data !== ' ') {
          matchFound = false
          break
        }
        matchFound = true
      }
    }
    if (matchFound === true) {
      return String.fromCodePoint('0'.codePointAt(0) + digit) 
    }
  }
  return '?'
}

export function convert (input) {
  const matrix = input.split('\n')
  const [nrows, ncols] = [matrix.length, matrix[0].length]
  const [nOutRows, nOutCols] = [Math.floor(nrows / 4), Math.floor(ncols / 3)]
  const result = Array.from({length: nOutRows})
  const buffer = []
  for (let i = 0; i < nOutRows; ++i) {
    buffer.length = 0
    for (let j = 0; j < nOutCols; ++j) {
      buffer.push(recognizeDigit(matrix, 4 * i, 3 * j))
    }
    result[i] = buffer.join('')
  }
  return result.join(',')
}